### ISSUE TITLE
afterAllArtifactBuild script fires before latest.yml is created

### ISSUE THREAD

**Issue by cameronsutter** [electron-userland/electron-builder#4446]
* **Version**: 21.0.1

* **Target**: mac: dmg, zip; win: nsis

In the afterAllArtifactBuild script, `latest.yml` (and `latest-mac.yml`) are not created yet or are not saved to the output directory yet.

mac CI logs output from afterAllArtifactBuild:
<img width="966" alt="mac_output" src="https://user-images.githubusercontent.com/780655/69369229-ed882780-0c58-11ea-990e-51920904bebd.png">

windows CI logs output from afterAllArtifactBuild:
<img width="856" alt="windows_output" src="https://user-images.githubusercontent.com/780655/69369245-f547cc00-0c58-11ea-8b5a-64552ab2ed2e.png">

The arrays starting on line 167 and 92 respectfully are all the files in the `outDir` folder.

Note: `BUILD TYPE trial` is something specific to my app

**Comment by kzimny** [comment]
Similar issue? #4435

**Comment by ShadixAced** [comment]
Maybe try to look at your folder when it's being built (try in your own PC, and then on CIs if it does change), or maybe it is already named like this.

**Comment by cameronsutter** [comment]
@kzimny it's not quite similar because that one deals with when the app is trying to update. My issue is during build.

@ShadixAced that's a good idea. I'll give that a try and see what's happening on my local computer

**Comment by oltreseba** [comment]
I'm having the same issue here. It really looks like afterAllArtifactBuild is called before latest-mac.yml is created. Therefore is quite a pain to setup a deploy system in the afterAllArtifactBuild.

I'm using version 22.2.0

**Comment by oltreseba** [comment]
Yes it is still relevant. latest-mac.yml (and alpha-mac.yml and beta-mac.yml if generateUpdatesFilesForAllChannels is set to true) is created after the event `afterAllArtifactBuild` is fired. It would make sense to fire the `afterAllArtifactBuild` event after all the files have been created.

**Comment by kroko** [comment]
Still relevant, because `latest-mac.yml` is actually not generated when version is valid semver alpha and `generateUpdatesFilesForAllChannels: true`.  
If `afterAllArtifactBuild` was predicatble a hook could be used, that would check if `latest-mac.yml` is generated, if not then create it by looking if there is `beta-mac.yml` or `alpha-mac.yml` and copy that as `latest-mac.yml`.

**Comment by asd281533890** [comment]
I solved this problem by using [chokidar](https://github.com/paulmillr/chokidar).

```JavaScript
// afterAllArtifactBuild.js
const chokidar = require('chokidar')
exports.default = async buildResult => {
  chokidar.watch('path/to/the/latest.yml').on('add', (event, path) => {
    // await do something
    process.exit(0)
  })
}
```

**Comment by asd281533890** [comment]
I solved this problem by using [chokidar](https://github.com/paulmillr/chokidar)

```JavaScript
// afterAllArtifactBuild.js
const chokidar = require('chokidar')
const fs = require('fs')
const path = require('path')
exports.default = async buildResult => {
  const filePath = path.join(buildResult.outDir, 'latest.yml')
  fs.access(filePath, fs.contants.R_OK, err => {
    const watcher = chokidar.watch(filePath)
    watcher.on(err ? 'add' : 'change', (event, path) => {
        // await do something
        watcher.unwatch(filePath)
        process.exit(0)
      })
  })
}
```

**Comment by panther7** [comment]
Still relevant in 23.0.0-alpha.3

**Comment by ArmelChesnais** [comment]
Still relevant in 23.0.2.
I'm trying to automate a modification to the yml file before it gets uploaded, and this is impossible directly with afterAllArtifactBuild, as the yml file doesn't exist when afterAllArtifactBuild is called. 
The above solution by asd seems like a possible hack/workaround, but doesn't seem like it should be the expected approach.

If afterAllArtifactBuild is meant to be used solely for the application artifacts and not the yml file, then a new hook along the lines of "beforeUpload" should be available.

**Comment by promotion-xu** [comment]
> I solved this problem by using [chokidar](https://github.com/paulmillr/chokidar)
> 
> ```js
> // afterAllArtifactBuild.js
> const chokidar = require('chokidar')
> const fs = require('fs')
> const path = require('path')
> exports.default = async buildResult => {
>   const filePath = path.join(buildResult.outDir, 'latest.yml')
>   fs.access(filePath, fs.contants.R_OK, err => {
>     const watcher = chokidar.watch(filePath)
>     watcher.on(err ? 'add' : 'change', (event, path) => {
>         // await do something
>         watcher.unwatch(filePath)
>         process.exit(0)
>       })
>   })
> }
> ```

fs.constants.R_OK is right

**Comment by MiguelSOliveira** [comment]
chokidar suggestion above works great for me.

Just did it in a slightly different way which was clearer, at least for me.

```
await chokidar
    .watch(buildResult.outDir, {
      awaitWriteFinish: {
        stabilityThreshold: 2000,
        pollInterval: 100
      }
    })
    .add('latest-mac.yml')
    .close()

... do whatever else
```

**Comment by MiguelSOliveira** [comment]
any news on this? can't seem to make it work with chokidar.

**Comment by mmaietta** [comment]
An alternative route is to use the JS API for electron-builder and then just `builder.build({...}).then(artifactPaths => yourDeployLogic())`
https://www.electron.build/programmatic-usage
